File: frontend/src/types/operations.ts
/** Generated schemas own stable DTOs; generic legacy/OPS dictionaries are explicit below. */
import type { components } from "@/api/generated";
type Schemas = components["schemas"];
export interface Page<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}
export interface LegacyPage<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}
export interface Choice {
  id: string;
  title?: string;
  name?: string;
  first_name?: string;
  last_name?: string;
  email?: string;
  email_opted_out?: boolean;
  do_not_contact?: boolean;
}
export type Conversation = Schemas["ConversationResponse"];
export type Participant = Schemas["ParticipantInput"];
export interface ConversationDetail {
  conversation: Conversation;
  participants: Participant[];
}
export interface Message {
  id: string;
  conversation_id: string;
  body: string;
  direction: string;
  status: string;
  sender: string | null;
  recipient: string | null;
  provider_message_id: string | null;
  attachments: { name: string; size?: number; content_type?: string }[];
  occurred_at: string;
  created_at: string;
}
export type Job = Schemas["JobResponse"];
export type ConversationInput = Schemas["ConversationInput"];
export interface Campaign {
  id: string;
  name: string;
  type: "broadcast" | "sequence" | "triggered";
  status: string;
  scheduled_at?: string | null;
  sender_name?: string;
  sender_email?: string;
  description?: string | null;
  audience?: { contact_ids: string[]; estimated_count: number } | null;
  templates?: { id: string; subject: string; step_order: number }[];
  sequences?: {
    id: string;
    delay_days: number;
    delay_hours: number;
    step_order: number;
  }[];
  stats?: Record<string, number> | null;
}
export interface CampaignCreate {
  name: string;
  type: Campaign["type"];
  sender_name: string;
  sender_email: string;
  description?: string;
}
export interface CampaignPlan {
  integration_id: string;
  contact_ids: string[];
  subject: string;
  body: string;
  steps: { subject: string; body: string; delay_hours: number }[];
  trigger_event?: string;
}
export interface Recipient {
  id: string;
  contact_id: string;
  contact_name?: string;
  email: string;
  status: string;
  sent_at?: string;
  error_message?: string;
  suppressed_reason?: string;
}
export interface CampaignStatistics {
  campaign_id: string;
  status: string;
  total_recipients: number;
  sent: number;
  queued: number;
  failed: number;
  suppressed: number;
  unknown?: number;
  opened: number;
  replied: number;
}
export const workflowTriggers = [
  "lead.created",
  "lead.updated",
  "deal.created",
  "deal.stage_changed",
  "task.overdue",
  "message.received",
  "campaign.completed",
] as const;
export const workflowTools = [
  "create_task",
  "update_lead",
  "update_deal",
  "add_note",
  "send_email",
  "assign_owner",
  "invoke_ai",
] as const;
export type WorkflowTool = (typeof workflowTools)[number];
export interface WorkflowCondition {
  field: "status" | "stage_id" | "ai_score" | "value" | "source" | "direction";
  operator: "eq" | "ne" | "gt" | "lt";
  value: string | number | boolean | null;
}
export interface WorkflowAction {
  tool: WorkflowTool;
  arguments: Record<string, string | number>;
}
export interface WorkflowInput {
  name: string;
  trigger: (typeof workflowTriggers)[number];
  conditions: WorkflowCondition[];
  actions: WorkflowAction[];
}
export interface Workflow {
  id: string;
  name: string;
  enabled: boolean;
  version: number;
}
export interface WorkflowDetail {
  id: string;
  enabled: boolean;
  version: number;
  definition: WorkflowInput | null;
}
export interface Execution {
  id: string;
  event_id: string;
  status: string;
  version: number;
  attempts: number;
  error_code: string | null;
  job_id?: string | null;
}
export interface Provider {
  provider: string;
  capabilities: string[];
  live_verified: boolean;
}
export interface Integration {
  id: string;
  name: string;
  type: string;
  status: string;
  sync_enabled: boolean;
  sync_frequency_minutes: number;
  last_sync_at: string | null;
  last_sync_status: string | null;
  records_synced: number;
}
export interface IntegrationStatus {
  id: string;
  status: string;
  capabilities: string[];
  expires_at: string | null;
  expired: boolean;
}
export interface SyncLog {
  id: string;
  status: string;
  started_at?: string;
  completed_at?: string;
  records_processed?: number;
  records_created?: number;
  records_updated?: number;
  error_message?: string;
}
export const aiCapabilities = [
  {
    value: "lead_score",
    label: "Lead scoring",
    entity: "leads",
    permission: "ai:score_leads",
  },
  {
    value: "lead_summary",
    label: "Lead summary",
    entity: "leads",
    permission: "ai:read_leads",
  },
  {
    value: "next_best_action",
    label: "Next best action",
    entity: "deals",
    permission: "ai:change_stage",
  },
  {
    value: "reply_analysis",
    label: "Reply analysis",
    entity: "messages",
    permission: "ai:read_replies",
  },
  {
    value: "campaign_draft",
    label: "Campaign drafting",
    entity: "campaigns",
    permission: "ai:draft_email",
  },
  {
    value: "proposal_draft",
    label: "Proposal drafting",
    entity: "deals",
    permission: "ai:create_proposal",
  },
] as const;
export type AICapability = (typeof aiCapabilities)[number]["value"];

File: app/schemas/operations.py
"""Stable, bounded request and response contracts for CRM operations."""

from datetime import datetime
from typing import Literal
from uuid import UUID
from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    EmailStr,
    AwareDatetime,
    model_validator,
)


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)


class Page(StrictModel):
    items: list[dict]
    total: int
    limit: int
    offset: int


class JobResponse(StrictModel):
    id: UUID
    kind: str
    status: str
    attempts: int
    result: dict
    error_code: str | None
    created_at: datetime


class ParticipantInput(StrictModel):
    address: str = Field(min_length=1, max_length=255)
    name: str | None = Field(default=None, max_length=255)


class ConversationInput(StrictModel):
    subject: str = Field(min_length=1, max_length=255)
    channel: Literal["email", "whatsapp", "social"] = "email"
    contact_id: UUID | None = None
    company_id: UUID | None = None
    lead_id: UUID | None = None
    integration_id: UUID | None = None
    participants: list[ParticipantInput] = Field(default_factory=list, max_length=20)


class ConversationResponse(StrictModel):
    id: UUID
    subject: str
    channel: str
    contact_id: UUID | None
    company_id: UUID | None
    lead_id: UUID | None
    integration_id: UUID | None
    unread_count: int
    last_message_at: datetime | None
    created_at: datetime


class Attachment(StrictModel):
    name: str = Field(min_length=1, max_length=255)
    content_type: str = Field(max_length=100)
    size: int = Field(ge=0, le=25_000_000)
    provider_attachment_id: str = Field(max_length=500)


class MessageInput(StrictModel):
    body: str = Field(min_length=1, max_length=100_000)
    recipient: EmailStr
    attachments: list[Attachment] = Field(default_factory=list, max_length=10)


class MessageResponse(StrictModel):
    id: UUID
    conversation_id: UUID
    body: str
    direction: str
    status: str
    sender: str | None
    recipient: str | None
    provider_message_id: str | None
    attachments: list[dict]
    occurred_at: datetime
    created_at: datetime


class CampaignPlan(StrictModel):
    integration_id: UUID
    contact_ids: list[UUID] = Field(default_factory=list, max_length=1000)
    company_id: UUID | None = None
    subject: str = Field(min_length=1, max_length=500)
    body: str = Field(min_length=1, max_length=100_000)
    steps: list[dict] = Field(default_factory=list, max_length=10)
    trigger_event: (
        Literal[
            "lead.created", "contact.created", "deal.stage_changed", "task.completed"
        ]
        | None
    ) = None


class ScheduleInput(StrictModel):
    scheduled_at: AwareDatetime


class WorkflowActionInput(StrictModel):
    tool: Literal[
        "create_task",
        "update_lead",
        "update_deal",
        "add_note",
        "send_email",
        "assign_owner",
        "invoke_ai",
    ]
    arguments: dict = Field(default_factory=dict)


class WorkflowConditionInput(StrictModel):
    field: Literal["status", "stage_id", "ai_score", "value", "source", "direction"]
    operator: Literal["eq", "ne", "gt", "lt"] = "eq"
    value: str | int | float | bool | None


class WorkflowInput(StrictModel):
    name: str = Field(min_length=1, max_length=255)
    trigger: Literal[
        "lead.created",
        "lead.updated",
        "deal.created",
        "deal.stage_changed",
        "task.overdue",
        "message.received",
        "campaign.completed",
    ]
    conditions: list[WorkflowConditionInput] = Field(
        default_factory=list, max_length=20
    )
    actions: list[WorkflowActionInput] = Field(min_length=1, max_length=10)


class ToggleInput(StrictModel):
    enabled: bool


class AIRequest(StrictModel):
    capability: Literal[
        "lead_score",
        "lead_summary",
        "next_best_action",
        "reply_analysis",
        "campaign_draft",
        "proposal_draft",
    ]
    entity_id: UUID


class AIBatchRequest(StrictModel):
    requests: list[AIRequest] = Field(min_length=1, max_length=25)


class ScoreOutput(StrictModel):
    score: int = Field(ge=0, le=100)
    confidence: float = Field(ge=0, le=1)
    reasons: list[str] = Field(min_length=1, max_length=10)


class SummaryOutput(StrictModel):
    summary: str = Field(min_length=1, max_length=5000)


class ActionOutput(StrictModel):
    action: str = Field(min_length=1, max_length=500)
    confidence: float = Field(ge=0, le=1)
    reasons: list[str] = Field(min_length=1, max_length=10)


class ReplyOutput(StrictModel):
    classification: Literal[
        "interested",
        "not_interested",
        "question",
        "objection",
        "meeting_request",
        "unsubscribe",
        "other",
    ]
    confidence: float = Field(ge=0, le=1)
    summary: str = Field(max_length=2000)


class DraftOutput(StrictModel):
    subject: str = Field(min_length=1, max_length=500)
    body: str = Field(min_length=1, max_length=20000)


class OAuthCallback(StrictModel):
    state: str = Field(min_length=32, max_length=200)
    code: str = Field(min_length=1, max_length=4000)


class ActivityInput(StrictModel):
    subject: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=10000)
    lead_id: UUID | None = None
    deal_id: UUID | None = None
    contact_id: UUID | None = None
    company_id: UUID | None = None

    @model_validator(mode="after")
    def relation(self):
        if not any((self.lead_id, self.deal_id, self.contact_id, self.company_id)):
            raise ValueError("An associated customer resource is required")
        return self


class ToolInput(StrictModel):
    arguments: dict = Field(default_factory=dict)


class CampaignCreateInput(StrictModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=10000)
    type: Literal["broadcast", "sequence", "triggered"] = "broadcast"
    sending_domain_id: UUID | None = None
    sender_name: str = Field(min_length=1, max_length=255)
    sender_email: EmailStr
    reply_to_email: EmailStr | None = None
    tags: list[str] = Field(default_factory=list, max_length=30)
